core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -173,16 +163,42 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * then the action will fail with some error.
    */
   final def raceBoth[E1 >: E, B](that: IO[E1, B]): IO[E1, Either[A, B]] =
-    raceWith(that)((a, fiber) => fiber.interrupt.const(Left(a)), (b, fiber) => fiber.interrupt.const(Right(b)))
+    raceWith(that)(_.join.map(Left(_)) <* _.interrupt, _.join.map(Right(_)) <* _.interrupt)
 
   /**
    * Races this action with the specified action, invoking the
    * specified finisher as soon as one value or the other has been computed.
    */
-  final def raceWith[E1 >: E, A1 >: A, B, C](
+  final def raceWith[E1, E2, B, C](
     that: IO[E1, B]
-  )(finishLeft: (A1, Fiber[E1, B]) => IO[E1, C], finishRight: (B, Fiber[E1, A1]) => IO[E1, C]): IO[E1, C] =
-    new IO.Race[E1, A1, B, C](self, that, finishLeft, finishRight)
+  )(
+    leftWins: (Fiber[E, A], Fiber[E1, B]) => IO[E2, C],
+    rightWins: (Fiber[E1, B], Fiber[E, A]) => IO[E2, C]
+  ): IO[E2, C] = {
+    def arbiter[E0, E1, A, B](
+      f: (Fiber[E0, A], Fiber[E1, B]) => IO[E2, C],
+      winner: Fiber[E0, A],
+      loser: Fiber[E1, B],
+      race: RefM[IO.Race],
```

Comment:
I would strongly recommend we not use `RefM` here, because `RefM` is tremendously slower than `Ref`. It does make things simpler but at too high a cost for such a fundamental operation as `raceWith`.

Reply:
Done, also much cleaner with `Ref`, the only requirement was that the state be updated atomically and `race.modify` is perfect for that.